Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -1674,6 +1678,116 @@ JS::GCPtr<CSSSupportsRule> Parser::convert_to_supports_rule(Rule& rule)
     return CSSSupportsRule::create(m_context.realm(), supports.release_nonnull(), rule_list);
 }
 
+JS::GCPtr<CSSPropertyRule> Parser::convert_to_property_rule(Rule& rule)
+{
+    // https://drafts.css-houdini.org/css-properties-values-api-1/#at-ruledef-property
+    // @property <custom-property-name> {
+    // <declaration-list>
+    // }
+
+    if (rule.prelude().is_empty()) {
+        dbgln_if(CSS_PARSER_DEBUG, "Failed to parse @property rule: Empty prelude.");
+        return {};
+    }
+
+    if (!rule.block()) {
+        dbgln_if(CSS_PARSER_DEBUG, "Failed to parse @property rule: No block.");
+        return {};
+    }
+
+    auto prelude_stream = TokenStream { rule.prelude() };
+    prelude_stream.discard_whitespace();
+    auto& token = prelude_stream.consume_a_token();
+    if (!token.is_token()) {
+        dbgln_if(CSS_PARSER_DEBUG, "CSSParser: @property has invalid prelude, prelude = {}; discarding.", rule.prelude());
+        return {};
+    }
+
+    auto name_token = token.token();
+    prelude_stream.discard_whitespace();
+
+    if (prelude_stream.has_next_token()) {
+        dbgln_if(CSS_PARSER_DEBUG, "CSSParser: @property has invalid prelude, prelude = {}; discarding.", rule.prelude());
+        return {};
+    }
+
+    if (!name_token.is(Token::Type::Ident)) {
+        dbgln_if(CSS_PARSER_DEBUG, "CSSParser: @property name is invalid: {}; discarding.", name_token.to_debug_string());
+        return {};
+    }
+
+    if (!name_token.ident().starts_with_bytes("--"sv)) {
+        dbgln_if(CSS_PARSER_DEBUG, "CSSParser: @property name doesn't start with '--': {}; discarding.", name_token.ident());
+        return {};
+    }
+
+    auto const& name = name_token.ident();
+
+    auto block_tokens = rule.block()->values();
+    auto block_stream = TokenStream { block_tokens };
+
+    auto list_of_declarations = parse_a_list_of_declarations(block_stream);
+
+    auto syntax_maybe = Optional<FlyString> {};
+    auto inherit_maybe = Optional<bool> {};
+    auto initial_value_maybe = Optional<FlyString> {};
+
+    for (auto const& declOrRule : list_of_declarations) {
+        if (!declOrRule.is_declaration()) {
+            dbgln_if(CSS_PARSER_DEBUG, "CSSParser: Only declarations allowed in @property block.; discarding.");
+            return {};
+        }
+        auto const& decl = declOrRule.declaration();
+
+        if (decl.name() == "syntax") {
+            if (syntax_maybe.has_value()) {
+                dbgln_if(CSS_PARSER_DEBUG, "CSSParser: Found multiple declarations of \"syntax\" in @property block; discarding.");
+                return {};
+            }
```

Comment:
Like with style properties, each declaration here (called a "descriptor") is allowed to appear multiple times, and we use whichever the last value was that we managed to parse. So it already having a value isn't an error. (Same for the others below.)

Reply:
TIL I knew that was the case for usual CSS descriptors. Dunno why I thought this was more magic